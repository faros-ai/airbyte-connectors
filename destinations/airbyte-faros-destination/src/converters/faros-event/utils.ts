import {ErrorObject} from 'ajv';
import {groupBy} from 'lodash';

import {Artifact, FarosPath, FarosTag, RunStep} from './types';

interface OrganizationKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeploymentKey {
  readonly uid: string;
  readonly source: string;
}

export interface DeployIds {
  readonly id: string;
  readonly source: string;
  readonly environment: string;
  readonly application: string;
}

export interface ArtifactKey {
  readonly uid: string;
  readonly repository: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface ArtifactIds {
  readonly id: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface CommitKey {
  readonly sha: string;
  readonly repository: {
    readonly name: string;
    readonly organization: OrganizationKey;
  };
}

export interface CommitIds {
  readonly sha: string;
  readonly repository: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildKey {
  readonly uid: string;
  readonly pipeline: {
    readonly uid: string;
    readonly organization: OrganizationKey;
  };
}

export interface BuildIds {
  readonly id: string;
  readonly pipeline: string;
  readonly organization: string;
  readonly source: string;
}

export interface BuildStepKey {
  readonly uid: string;
  readonly build: BuildKey;
}

export interface AgentKey {
  readonly uid: string;
  readonly source: string;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveCommitIds(commit: any): CommitIds {
  if (commit.uri) {
    const parsedUri = parseUri(commit.uri);
    return {
      sha: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    sha: commit.sha,
    repository: commit.repository.toLowerCase(),
    organization: commit.organization.toLowerCase(),
    source: commit.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeCommitKey(commit: any): CommitKey {
  const commitIds = resolveCommitIds(commit);
  return {
    sha: commitIds.sha,
    repository: {
      name: commitIds.repository,
      organization: {
        uid: commitIds.organization,
        source: commitIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveBuildIds(run: any): BuildIds {
  if (run.uri) {
    const parsedUri = parseUri(run.uri);
    return {
      id: parsedUri[3],
      pipeline: parsedUri[2],
      organization: parsedUri[1],
      source: parsedUri[0],
    };
  }
  return {
    id: run.id,
    pipeline: run.pipeline,
    organization: run.organization,
    source: run.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeBuildKey(run: any): BuildKey {
  const buildIds = resolveBuildIds(run);
  return {
    uid: buildIds.id,
    pipeline: {
      uid: buildIds.pipeline,
      organization: {
        uid: buildIds.organization,
        source: buildIds.source,
      },
    },
  };
}

export function makeBuildStepKey(step: RunStep, build: BuildKey): BuildStepKey {
  return {
    uid: step.id,
    build,
  };
}

export function resolveDeploymentIds(deploy: any): DeployIds {
  let deployIds;
  if (deploy.uri) {
    const parsedUri = parseUri(deploy.uri);
    deployIds = {
      id: parsedUri[3],
      environment: parsedUri[2],
      application: parsedUri[1],
      source: parsedUri[0],
    };
  } else {
    deployIds = {
      id: deploy.id,
      environment: deploy.environment,
      application: deploy.application,
      source: deploy.source,
    };
  }
  return deployIds;
}

function enumToMap(enumValues: string[]): Map<string, string> {
  return new Map(enumValues.map((value) => [value.toLowerCase(), value]));
}

function resolveEnum(
  enumValue: Map<string, string>,
  field: string,
  fieldDetail?: string | null
): {category: string; detail?: string | null} {
  const lowerField = field.toLowerCase();
  const value = enumValue.get(lowerField);
  if (value) {
    return {category: value, detail: fieldDetail};
  }
  return {
    category: 'Custom',
    detail: field.concat(fieldDetail ? ` - ${fieldDetail}` : ''),
  };
}

const deploymentEnvironments = enumToMap([
  'Prod',
  'Staging',
  'QA',
  'Dev',
  'Sandbox',
  'Canary',
  'Custom',
]);

export function resolveDeploymentEnvironment(
  environment: string,
  environmentDetails?: string | null
): {category: string; detail?: string | null} {
  return resolveEnum(deploymentEnvironments, environment, environmentDetails);
}

const deploymentStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'RolledBack',
  'Success',
]);

export function resolveDeploymentStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(deploymentStatuses, status, statusDetails);
}

const runStatuses = enumToMap([
  'Canceled',
  'Custom',
  'Failed',
  'Queued',
  'Running',
  'Success',
  'Unknown',
]);

export function resolveRunStatus(
  status?: string,
  statusDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!status) {
    return undefined;
  }
  return resolveEnum(runStatuses, status, statusDetails);
}

const runStepTypes = enumToMap(['Custom', 'Manual', 'Script']);

export function resolveRunStepType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(runStepTypes, type, typeDetails);
}

const computeAgentOS = enumToMap(['Linux', 'MacOS', 'Windows', 'Custom']);

export function resolveComputeAgentOS(
  os?: string,
  osDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!os) {
    return undefined;
  }
  return resolveEnum(computeAgentOS, os, osDetails);
}

const metricDefinitionTypes = enumToMap([
  'Numeric',
  'String',
  'Date',
  'Timestamp',
  'Duration',
  'Percentage',
  'Custom',
]);

export function resolveMetricDefinitionType(
  type?: string,
  typeDetails?: string | null
): {category: string; detail?: string | null} | undefined {
  if (!type) {
    return undefined;
  }
  return resolveEnum(metricDefinitionTypes, type, typeDetails);
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function resolveArtifactIds(artifact: any): ArtifactIds {
  if (artifact.uri) {
    const parsedUri = parseUri(artifact.uri);
    return {
      id: parsedUri[3],
      repository: parsedUri[2].toLowerCase(),
      organization: parsedUri[1].toLowerCase(),
      source: parsedUri[0],
    };
  }
  return {
    id: artifact.id,
    repository: artifact.repository.toLowerCase(),
    organization: artifact.organization.toLowerCase(),
    source: artifact.source,
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeArtifactKey(artifact: any): ArtifactKey {
  const artifactIds = resolveArtifactIds(artifact);
  return {
    uid: artifactIds.id,
    repository: {
      uid: artifactIds.repository.toLowerCase(),
      organization: {
        uid: artifactIds.organization.toLowerCase(),
        source: artifactIds.source,
      },
    },
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function makeDummyArtifact(commit: any): Artifact {
  const commitKey = makeCommitKey(commit);
  return {
    id: commitKey.sha,
    repository: commitKey.repository.name,
    organization: commitKey.repository.organization.uid,
    source: commitKey.repository.organization.source,
  };
}

// Convert ISO-8601 string, number string, number, or 'Now' into milliseconds
export function getTime(date?: string | number): number | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return Date.now();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).getTime();
      }
    }
    return new Date(date).getTime();
  }
  return undefined;
}

// Convert ISO-8601 string, number string, number, or 'Now' into ISO-8601 string
export function toDate(date?: string | number): string | undefined {
  if (date) {
    if (typeof date === 'string') {
      if (date === 'Now') {
        return new Date(Date.now()).toISOString();
      }
      const dateInt = Number(date);
      if (!isNaN(dateInt)) {
        return new Date(dateInt).toISOString();
      }
    }
    return new Date(date).toISOString();
  }
  return undefined;
}

export function parseTags(tags?: FarosTag[] | string): FarosTag[] | undefined {
  const parser = (input: string): FarosTag[] => {
    const resTags: FarosTag[] = [];
    const keyValues = input.split(',');
    for (const keyValue of keyValues) {
      const splitTag = keyValue.split(':');
      resTags.push({key: splitTag[0], value: splitTag[1]});
    }
    return resTags;
  };
  return parseArray(parser, tags);
}

export function parsePaths(
  paths?: FarosPath[] | string
): FarosPath[] | undefined {
  const parser = (input: string): FarosPath[] => {
    return input.split(',').map<FarosPath>((path) => {
      return {path};
    });
  };
  return parseArray(parser, paths);
}

export function parsePullRequestNumber(
  number: string | number
): number | undefined {
  const prNum = typeof number === 'string' ? Number(number) : number;
  return isNaN(prNum) ? undefined : prNum;
}

export function parseUri(uri: string): string[] {
  const components: string[] = [];
  const split1 = uri.split('://');
  components.push(split1[0]);
  split1[1].split('/').forEach((value) => components.push(value));
  return components;
}

export function parseStringArray(
  array?: string[] | string,
  separator = ','
): string[] | undefined {
  return parseArray((s) => s.split(separator), array);
}

function parseArray<T>(
  parser: (input: string) => T[],
  input?: T[] | string
): T[] | undefined {
  if (typeof input === 'string') {
    return parser(input);
  }
  return input;
}

export interface EventValidationError {
  readonly message: string;
  readonly details?: ReadonlyArray<string>;
  readonly allowedValues?: ReadonlyArray<string>;
  readonly additionalProperties?: ReadonlyArray<string>;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function removeUndefinedProperties(object: any): any {
  const result = {...object};
  Object.keys(result).forEach(
    (key) => result[key] === undefined && delete result[key]
  );
  return result;
}

export function formatErrors(
  errors?: ErrorObject[] | null,
  rootObject?: string
): ReadonlyArray<EventValidationError> {
  if (!errors) {
    return [];
  }

  const formattedErrors: EventValidationError[] = [];
  const groupedErrors = groupBy<ErrorObject>(errors, (err) => {
    return formatErrorDetailsMessage(err, rootObject);
  });

  for (const [message, errs] of Object.entries(groupedErrors)) {
    const details = errs.flatMap(
      (err) =>
        err.params?.errors?.map((v: any) =>
          formatErrorDetailsMessage(v, rootObject)
        ) ?? []
    );
    const additionalProperties = errs.flatMap(
      (err) => err.params?.additionalProperty ?? []
    );
    formattedErrors.push({
      message,
      details: details.length > 0 ? details : undefined,
      allowedValues: getAllowedValues(errs),
      additionalProperties:
        additionalProperties.length > 0 ? additionalProperties : undefined,
    });
  }
  return formattedErrors;
}

function formatErrorDetailsMessage(
  error: ErrorObject,
  rootObject = 'data'
): string {
  const prefix = error?.instancePath
    ? `${rootObject}${error.instancePath.replace(/\//g, '.')} `
    : `${rootObject} `;
  return `${prefix}${error.message}`;
}

function getAllowedValues(errors: ErrorObject[]): string[] | undefined {
  let result;
  if (errors?.length > 0) {
    result = errors[0].params?.allowedValues;
    if (!result && errors[0].params?.errors?.length > 0) {
      result = errors[0].params?.errors[0].params?.allowedValues;
    }
  }
  return result;
}

// Max length for free-form description text fields such as issue body
const MAX_TEXT_LENGTH = 1000;
export function cleanAndTruncate(
  str?: string,
  maxLength?: number
): string | null | undefined {
  if (!str) {
    return str;
  }
  const length = maxLength ?? MAX_TEXT_LENGTH;
  let result: string;
  if (str.length <= length) {
    result = str;
  } else {
    // If the last character is part of a unicode surrogate pair, include the next character
    const lastChar = str.codePointAt(length - 1) ?? 0;
    result =
      lastChar > 65535
        ? str.substring(0, length + 1)
        : str.substring(0, length);
  }
  // eslint-disable-next-line no-control-regex
  return result.replace(/\u0000/g, '');
}
