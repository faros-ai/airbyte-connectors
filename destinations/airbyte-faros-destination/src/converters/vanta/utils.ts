import fs from 'fs-extra';
import path from 'path';

// Function to find a specific parent directory
function findParentDirectory(
  currentDir: string,
  targetDirName: string
): string | null {
  let currentPath = currentDir;
  while (currentPath !== '/') {
    if (path.basename(currentPath) === targetDirName) {
      return currentPath;
    }
    currentPath = path.dirname(currentPath);
  }
  return null;
}

export function getQueryFromName(name: string): string {
  const crt_dir = __dirname;
  // We go back to the part of dir which ends with name:
  const targetDirPath = findParentDirectory(
    crt_dir,
    'airbyte-faros-destination'
  );
  if (!targetDirPath) {
    throw new Error('airbyte-faros-destination directory not found');
  }

  const gql_dir = path.join(targetDirPath, 'resources');

  const fn = `${name}.gql`;
  // Check if fn is in gql_dir:
  if (!fs.existsSync(`${gql_dir}/${fn}`)) {
    throw new Error(`Query ${fn} not found in ${gql_dir}`);
  }
  // We access the directory resources/gql_queries to get the query
  return fs.readFileSync(`${gql_dir}/${fn}`, 'utf8');
}

export function looksLikeGitCommitSha(sha: string): boolean {
  return /^[a-f0-9]{40}$/i.test(sha);
}
