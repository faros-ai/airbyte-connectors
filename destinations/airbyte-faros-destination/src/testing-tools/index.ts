export * from './cli';
export * from './testing-tools';
export * from './utils';
