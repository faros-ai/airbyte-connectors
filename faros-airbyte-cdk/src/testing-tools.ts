import fs from 'fs';
import {Dictionary} from 'ts-essentials';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
} from '.';

export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}
