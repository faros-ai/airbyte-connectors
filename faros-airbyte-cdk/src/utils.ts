import fastRedact from 'fast-redact';
import fs from 'fs';
import traverse from 'json-schema-traverse';
import _, {cloneDeep} from 'lodash';
import path from 'path';

import {AirbyteConfig, AirbyteSpec} from './protocol';

export const PACKAGE_ROOT = path.join(__dirname, '..');

const packageInfo = JSON.parse(
  fs.readFileSync(path.join(PACKAGE_ROOT, 'package.json'), 'utf8')
);

export const PACKAGE_VERSION = packageInfo.version;

export function pathsToRedact(spec: AirbyteSpec): string[] {
  const paths = [];
  traverse(spec.spec.connectionSpecification ?? {}, {
    cb: (schema, pointer) => {
      if (schema.airbyte_secret) {
        paths.push(toPath(pointer));
      }
    },
  });
  return paths;
}

/** Redact config of all secret values based on the provided specification */
export function redactConfig(
  config: AirbyteConfig,
  spec: AirbyteSpec
): AirbyteConfig {
  const redact = fastRedact({
    paths: pathsToRedact(spec),
    censor: 'REDACTED',
    serialize: false,
  });
  return redact(_.cloneDeep(config)) as AirbyteConfig;
}

export function redactConfigAsString(
  config: AirbyteConfig,
  spec: AirbyteSpec
): string {
  const redact = fastRedact({paths: pathsToRedact(spec), censor: 'REDACTED'});
  return `${redact(config)}`;
}

function toPath(pointer: string): string {
  return pointer
    .replace(/\/oneOf\/\d+/g, '')
    .replace(/\/items(\/?)/g, '[*]$1')
    .split('/properties/')
    .filter((s) => s)
    .join('.');
}

/**
 * Sets all undefined values with defaults from spec to their default value.
 * The changes are made on a copy of the input.
 * */
export function withDefaults<Config extends AirbyteConfig>(
  config: Config,
  spec: AirbyteSpec
): Config {
  const defaultsByPath = new Map<string, any>();
  traverse(spec.spec.connectionSpecification ?? {}, {
    cb: (schema, pointer) => {
      if (schema.default) {
        defaultsByPath.set(toPath(pointer), schema.default);
      }
    },
  });
  // create a copy of input and apply defaults
  const result = _.cloneDeep(config);
  for (const [path, defaultValue] of defaultsByPath) {
    if (_.get(result, path) === undefined) {
      _.set(result, path, defaultValue);
    }
  }
  return result;
}

/** Convert a value to Date */
export function toDate(
  val: Date | string | number | undefined
): Date | undefined {
  if (typeof val === 'number') {
    return new Date(val);
  }
  if (!val) {
    return undefined;
  }
  return new Date(val);
}

export function base64Encode(str: string): string {
  return Buffer.from(str, 'binary').toString('base64');
}

type Properties = {[propName: string]: SpecProperty};

interface SpecObject {
  title: string;
  type: string;
  required?: string[];
  properties: Properties;
}

interface SpecProperty {
  type: string;
  order?: number;
  const?: string;
  title?: string;
  description?: string;
  examples?: any[];
  items?: {type: string; title?: string; properties?: Properties};
  oneOf?: SpecObject[];
  properties?: Properties;
}

export function minimizeSpec(airbyteSpec: AirbyteSpec): AirbyteSpec {
  const spec = cloneDeep(airbyteSpec.spec);
  for (const prop of Object.values(
    spec.connectionSpecification?.properties ?? {}
  )) {
    minimizeSpecProperty(prop as SpecProperty);
  }
  return new AirbyteSpec(spec);
}

function minimizeSpecProperty(prop: SpecProperty): void {
  prop.title = undefined;
  prop.description = undefined;
  prop.examples = undefined;
  prop.order = undefined;
  for (const object of prop.oneOf ?? []) {
    minimizeSpecObject(object);
  }
  for (const item of Object.values(prop.items?.properties ?? {})) {
    minimizeSpecProperty(item);
  }
  for (const property of Object.values(prop.properties ?? {})) {
    minimizeSpecProperty(property);
  }
}

function minimizeSpecObject(config: SpecObject): void {
  for (const prop of Object.values(config.properties)) {
    minimizeSpecProperty(prop);
  }
}
