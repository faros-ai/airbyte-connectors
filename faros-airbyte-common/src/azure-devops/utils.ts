import {User} from './types';

/**
 * Gets a unique identifier for an Azure DevOps user.
 * Tries principalName first, then falls back to uniqueName.
 * This handles differences between Azure DevOps Cloud and Server.
 *
 * @param user - The Azure DevOps user object
 * @returns The user's unique identifier, or undefined if neither field exists
 */
export function getUserIdentifier(user: User): string | undefined {
  if ('principalName' in user && user.principalName) {
    return user.principalName;
  }
  if ('uniqueName' in user && user.uniqueName) {
    return user.uniqueName;
  }
  return undefined;
}
