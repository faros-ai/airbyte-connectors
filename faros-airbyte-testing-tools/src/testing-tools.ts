import fs from 'fs';
import {Mockttp} from 'mockttp';
import pino from 'pino';
import tmp from 'tmp-promise';
import {Dictionary} from 'ts-essentials';
import util from 'util';

import {
  AirbyteConfig,
  AirbyteConfiguredCatalog,
  AirbyteMessageType,
  AirbyteRecord,
  AirbyteSourceBase,
  AirbyteState,
  AirbyteStateMessage,
} from 'faros-airbyte-cdk';
import {Data} from 'faros-airbyte-cdk/lib/utils';
import {Edition, InvalidRecordStrategy} from 'airbyte-faros-destination/lib/common/types';

const TEST_SOURCE_ID = 'mytestsource';

tmp.setGracefulCleanup();

/**
 * Read a test resource by name
 */
export function readTestResourceFile(fileName: string): string {
  return fs.readFileSync(`test/resources/${fileName}`, 'utf8');
}

const writeFile = util.promisify(fs.write);

/**
 * Parse a test resource into JSON
 */
export function readTestResourceAsJSON(fileName: string): any {
  return JSON.parse(readTestResourceFile(fileName));
}

/**
 * Creates a temporary file
 * @return path to the temporary file
 */
export async function tempFile(data: string, postfix: string): Promise<string> {
  const file = await tmp.file({postfix});
  await writeFile(file.fd, data, null, 'utf-8');
  return file.path;
}

export function readTestFile(fileName: string): string {
  return fs.readFileSync(`test_files/${fileName}`, 'utf8');
}

/**
 * Parse a test file into JSON
 */
export function readTestFileAsJSON(fileName: string): any {
  return JSON.parse(readTestFile(fileName));
}

export function readResourceFile(fileName: string): string {
  return fs.readFileSync(`resources/${fileName}`, 'utf8');
}

/**
 * Parse a resource into JSON
 */
export function readResourceAsJSON(fileName: string): any {
  return JSON.parse(readResourceFile(fileName));
}

export interface TempConfigOptions {
  readonly api_url: string;
  readonly invalid_record_strategy?: InvalidRecordStrategy;
  readonly edition?: Edition;
  readonly edition_configs?: Dictionary<any>;
  readonly source_specific_configs?: Dictionary<any>;
  readonly replace_origin_map?: Dictionary<any>;
  readonly exclude_fields_map?: Dictionary<any>;
  readonly log_records?: boolean;
}

/**
 * Creates a temporary file with testing configuration
 * @return path to the temporary config file to delete
 */
export async function tempConfig(options: TempConfigOptions): Promise<string> {
  const conf = getConf(options);

  return tempFile(JSON.stringify(conf), '.json');
}

export function getConf(options: TempConfigOptions): any {
  const {
    api_url,
    invalid_record_strategy = InvalidRecordStrategy.FAIL,
    edition = Edition.CLOUD,
    edition_configs = {},
    source_specific_configs = {},
    replace_origin_map = {},
    exclude_fields_map = {},
    log_records = false,
  } = options;

  const edition_configs_defaults =
    edition === Edition.CLOUD
      ? {
          edition,
          api_url,
          api_key: 'test-api-key',
          graph: 'test-graph',
        }
      : {
          edition,
          hasura_url: api_url,
          segment_user_id: 'bacaf6e6-41d8-4102-a3a4-5d28100e642f',
          segment_test_host: api_url,
        };

  const conf = {
    edition_configs: {...edition_configs_defaults, ...edition_configs},
    invalid_record_strategy,
    origin: 'test-origin',
    jsonata_destination_models: ['generic_Record'],
    jsonata_expression: `
    data.{
      "model": "generic_Record",
      "record": {
        "uid": foo
      }
    }`,
    source_specific_configs,
    replace_origin_map: JSON.stringify(replace_origin_map),
    exclude_fields_map: JSON.stringify(exclude_fields_map),
    faros_source_id: TEST_SOURCE_ID,
    log_records,
  };

  return conf;
}

export function sourceSpecificTempConfig(
  url: string,
  source_specific_configs: Dictionary<any>
): Promise<string> {
  return tempConfig({api_url: url, source_specific_configs});
}

export async function initMockttp(mockttp: Mockttp): Promise<void> {
  await mockttp.start({startPort: 30000, endPort: 50000});

  await mockttp
    .forGet('/users/me')
    .once()
    .thenReply(200, JSON.stringify({tenantId: '1'}));

  await mockttp
    .forGet('/graphs/test-graph/statistics')
    .once()
    .thenReply(200, JSON.stringify({}));

  await mockttp.forGet('/healthz').once().thenReply(200, JSON.stringify({}));

  await mockttp
    .forGet(`/accounts/${TEST_SOURCE_ID}`)
    .once()
    .thenReply(200, JSON.stringify({account: {accountId: TEST_SOURCE_ID}}));

  const mockSyncResult = {
    sync: {
      syncId: '1',
      logId: '1',
      startedAt: new Date().toISOString(),
      status: 'running',
    },
  };
  await mockttp
    .forPut(`/accounts/${TEST_SOURCE_ID}/syncs`)
    .once()
    .thenReply(200, JSON.stringify(mockSyncResult));

  await mockttp
    .forPatch(`/accounts/${TEST_SOURCE_ID}/syncs/1`)
    .once()
    .thenReply(200, JSON.stringify(mockSyncResult));
}

export function testLogger(name = 'test'): pino.Logger {
  return pino({
    name,
    level: process.env.LOG_LEVEL ?? 'info',
    transport:
      process.env.LOG_LEVEL?.toLowerCase() === 'debug'
        ? {target: 'pino-pretty', options: {levelFirst: true}}
        : undefined,
  });
}

export interface SourceCheckTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath?: string | AirbyteConfig;
}

export const sourceCheckTest = async (
  options: SourceCheckTestOptions
): Promise<void> => {
  const {source, configOrPath} = options;
  const config = resolveInput(configOrPath);
  expect(await source.check(config)).toMatchSnapshot();
};

export interface SourceReadTestOptions {
  source: AirbyteSourceBase<AirbyteConfig>;
  configOrPath: string | AirbyteConfig;
  catalogOrPath: string | AirbyteConfiguredCatalog;
  stateOrPath?: string | AirbyteState;
  onBeforeReadResultConsumer?: (res: {
    config: AirbyteConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }) => void;
  checkRecordsData?: (records: ReadonlyArray<Dictionary<any>>) => void;
  checkFinalState?: (state: Dictionary<any>) => void;
}

export const sourceReadTest = async (
  options: SourceReadTestOptions
): Promise<void> => {
  const {
    source,
    configOrPath,
    catalogOrPath,
    stateOrPath = undefined,
    onBeforeReadResultConsumer = undefined,
    checkRecordsData = undefined,
    checkFinalState = undefined,
  } = options;
  const config = resolveInput(configOrPath);
  const catalog = resolveInput(catalogOrPath);
  const state = stateOrPath ? resolveInput(stateOrPath) : {};
  const res = await source.onBeforeRead(config, catalog, state);
  if (onBeforeReadResultConsumer) {
    onBeforeReadResultConsumer(res);
  }
  const generator = source.read(res.config, res.config, res.catalog, res.state);
  const records: Dictionary<any>[] = [];
  let finalState: Dictionary<any>;
  for await (const message of generator) {
    if (message.type === AirbyteMessageType.RECORD) {
      records.push((message as AirbyteRecord).record.data);
    } else if (message.type === AirbyteMessageType.STATE) {
      finalState = (message as AirbyteStateMessage).state.data;
    }
  }
  if (checkRecordsData) {
    checkRecordsData(records);
  }
  if (checkFinalState) {
    checkFinalState(Data.decompress(finalState));
  }
};

export const sourceSchemaTest = (
  source: AirbyteSourceBase<AirbyteConfig>,
  config: AirbyteConfig
): void => {
  const streams = source.streams(config);

  const validateFieldInSchema = (
    field: string | string[],
    schema: Dictionary<any>
  ): void => {
    if (Array.isArray(field)) {
      let nestedField = schema;
      for (const subField of field) {
        expect(nestedField).toHaveProperty(subField);
        nestedField = nestedField[subField].properties;
      }
    } else {
      expect(schema).toHaveProperty(field);
    }
  };

  for (const stream of streams) {
    const jsonSchema = stream.getJsonSchema().properties;
    const primaryKey = stream.primaryKey;
    const cursorField = stream.cursorField;

    if (primaryKey) {
      if (Array.isArray(primaryKey)) {
        for (const key of primaryKey) {
          validateFieldInSchema(key, jsonSchema);
        }
      } else {
        validateFieldInSchema(primaryKey, jsonSchema);
      }
    }

    validateFieldInSchema(cursorField, jsonSchema);
  }
};

function resolveInput<T>(inputOrPath: string | T): T {
  return typeof inputOrPath === 'string'
    ? readTestResourceAsJSON(inputOrPath)
    : inputOrPath;
}

export const customStreamsTest = async (
  source: AirbyteSourceBase<AirbyteConfig>,
  config: AirbyteConfig,
  allStreams: ReadonlyArray<string>,
  selectedStreams?: ReadonlyArray<string>
): Promise<void> => {
  const {catalog} = await source.onBeforeRead(
    {...config, run_mode: 'Custom', custom_streams: selectedStreams},
    {streams: allStreams.map((name) => ({stream: {name}}))} as any
  );
  expect(catalog.streams).toHaveLength((selectedStreams ?? allStreams).length);
  expect(catalog.streams.map((s) => s.stream.name)).toEqual(
    expect.arrayContaining(selectedStreams ?? allStreams)
  );
};
