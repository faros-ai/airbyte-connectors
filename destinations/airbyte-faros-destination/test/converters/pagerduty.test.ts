import {getLocal} from 'mockttp';

import {initMockttp, sourceSpecificTempConfig} from '../testing-tools';
import {destinationWriteTest} from './utils';

describe('pagerduty', () => {
  const mockttp = getLocal({debug: false, recordTraffic: false});
  let configPath: string;

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await sourceSpecificTempConfig(mockttp.url, {
      pagerduty: {associate_applications_to_teams: true},
    });
    await mockttp
      .forPost('/graphs/test-graph/graphql')
      .once()
      .thenReply(
        200,
        JSON.stringify({
          data: {
            org: {teams: {edges: [{node: {uid: 'eng', name: 'Engineering'}}]}},
          },
        })
      );
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    await destinationWriteTest({
      configPath,
      catalogPath: 'test/resources/pagerduty/catalog.json',
      inputRecordsPath: 'pagerduty/all-streams.log',
    });
  });
});
