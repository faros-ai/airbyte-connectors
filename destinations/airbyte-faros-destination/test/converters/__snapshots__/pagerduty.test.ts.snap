// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagerduty process records from all streams 1`] = `
Array [
  "Processed 8 records",
  "Processed records by stream: {\\\\\\"mytestsource__pagerduty__incident_log_entries\\\\\\":3,\\\\\\"mytestsource__pagerduty__incidents\\\\\\":3,\\\\\\"mytestsource__pagerduty__services\\\\\\":1,\\\\\\"mytestsource__pagerduty__users\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 16 records",
  "Would write records by model: {\\\\\\"compute_Application\\\\\\":2,\\\\\\"ims_Incident\\\\\\":3,\\\\\\"ims_IncidentApplicationImpact\\\\\\":3,\\\\\\"ims_IncidentAssignment\\\\\\":3,\\\\\\"ims_IncidentEvent\\\\\\":3,\\\\\\"ims_User\\\\\\":1,\\\\\\"org_ApplicationOwnership\\\\\\":1}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
